src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java

```diff
@@ -576,31 +592,74 @@ private static String getStringInsideTag(String result, DetailNode detailNode) {
     }
 
     /**
-     * Finds and returns first sentence.
+     * Finds and returns the first sentence.
      *
-     * @param ast Javadoc root node.
-     * @return first sentence.
+     * @param ast The Javadoc root node.
+     * @param period The configured period symbol.
+     * @return The first sentence up to and excluding the period, if a sentence ending was found.
      */
-    private static String getFirstSentence(DetailNode ast) {
-        final StringBuilder result = new StringBuilder(256);
-        final String periodSuffix = DEFAULT_PERIOD + ' ';
-        for (DetailNode child : ast.getChildren()) {
-            final String text;
-            if (child.getChildren().length == 0) {
-                text = child.getText();
+    private static Optional<String> getFirstSentence(DetailNode ast, String period) {
+        final List<String> sentenceParts = new ArrayList<>();
+        String sentence = null;
+        for (String text : (Iterable<String>) streamTextParts(ast)::iterator) {
+            final String sentenceEnding = findSentenceEnding(text, period).orElse(null);
+            if (sentenceEnding != null) {
+                sentenceParts.add(sentenceEnding);
+                sentence = String.join("", sentenceParts);
+                break;
             }
             else {
-                text = getFirstSentence(child);
+                sentenceParts.add(text);
             }
+        }
+        return Optional.ofNullable(sentence);
+    }
+
+    /**
+     * Streams through all the text under the given node.
+     *
+     * @param node The Javadoc node to examine.
+     * @return All the text in all nodes that have no child nodes.
+     */
+    private static Stream<String> streamTextParts(DetailNode node) {
+        final Stream<String> stream;
+        if (node.getChildren().length == 0) {
+            stream = Stream.of(node.getText());
+        }
+        else {
+            stream = Stream.of(node.getChildren())
+                .flatMap(SummaryJavadocCheck::streamTextParts);
+        }
+        return stream;
+    }
 
-            if (text.contains(periodSuffix)) {
-                result.append(text, 0, text.indexOf(periodSuffix) + 1);
+    /**
+     * Finds the end of a sentence. If a sentence ending period was found, returns the whole string
+     * up to and excluding that period. The end of sentence detection here could be replaced in the
+     * future by Java's built-in BreakIterator class.
+     *
+     * @param text The string to search.
+     * @param period The period character to find.
+     * @return The string up to and excluding the period, if one was found.
+     */
+    private static Optional<String> findSentenceEnding(String text, String period) {
```

Comment:
Checker fails if I make it return null. Will it pass if I annotate it with @Nullable or should I put it back to use an Optional?

Reply:
>Will it pass if I annotate it with @Nullable or should I put it back to use an Optional?

You can try to use this annotation, yes. However, if this makes pitest fail, we can just use Optional.

Reply:
It was checkerframefork failure, not a pitest.

Reply:
We should return Optional and do IF present Else ..

Reply:
Done

Reply:
I can't find a better way to combine the optional with the break statement in `getFirstSentence`.

If I try to use `.isPresent` or `.isEmpty` combined with a `.get`, it tells me to use `.orElseThrow` instead. That looks confusing, because it implies there's some way the optional could still be empty after the isPresent/isEmpty check.

We can't use `.ifPresent` or `.ifPresentOrElse` because you can't put the break statement inside the lambda.

Can you think of another way to do it? I think I'd be fine with leaving it as `.orElse(null)`.

Reply:
To be 100% clear: I am fine to go back to https://github.com/checkstyle/checkstyle/commit/b48508bd33b64496067bdf98939e5d830bda9a4c if removing Optional use and adding `@Nullable` annotation does not make checker happy.

Reply:
I've just pushed a version with the nullable annotation.

Reply:
@patchwork01 take a look at other `@Nullable` usage, we should use `javax.annotation.Nullable` I believe. Also, please place annotations above the method definition.